clang/lib/AST/Expr.cpp

```diff
@@ -568,6 +568,78 @@ std::string SYCLUniqueStableNameExpr::ComputeName(ASTContext &Context,
+                                               SourceLocation LParen,
+                                               SourceLocation RParen,
+                                               QualType ResultTy, Expr *E)
+    : Expr(SYCLUniqueStableIdExprClass, ResultTy, VK_RValue, OK_Ordinary),
+      OpLoc(OpLoc), LParen(LParen), RParen(RParen), DRE(E) {
+  setDependence(computeDependence(this));
+}
+
+SYCLUniqueStableIdExpr *SYCLUniqueStableIdExpr::Create(const ASTContext &Ctx,
+                                                       SourceLocation OpLoc,
+                                                       SourceLocation LParen,
+                                                       SourceLocation RParen,
+                                                       Expr *E) {
+  QualType ResultTy = Ctx.getPointerType(Ctx.CharTy.withConst());
+  return new (Ctx) SYCLUniqueStableIdExpr(OpLoc, LParen, RParen, ResultTy, E);
+}
+
+SYCLUniqueStableIdExpr *
+SYCLUniqueStableIdExpr::CreateEmpty(const ASTContext &Ctx) {
+  QualType ResultTy = Ctx.getPointerType(Ctx.CharTy.withConst());
+  return new (Ctx) SYCLUniqueStableIdExpr(EmptyShell(), ResultTy);
+}
+
+std::string SYCLUniqueStableIdExpr::ComputeName(ASTContext &Context) const {
+  assert(!isInstantiationDependent() &&
+         "Can't compute name of uninstantiated value");
+
+  auto *DR = cast<DeclRefExpr>(getExpr());
+  auto *VD = cast<VarDecl>(DR->getDecl());
+
+  return ComputeName(Context, VD);
+}
+
+std::string SYCLUniqueStableIdExpr::ComputeName(ASTContext &Context,
+                                                const VarDecl *VD) {
+  auto MangleCallback = [](ASTContext &Ctx,
+                           const NamedDecl *ND) -> llvm::Optional<unsigned> {
+    // This replaces the 'lambda number' in the mangling with a unique number
+    // based on its order in the declaration.  To provide some level of visual
+    // notability (actual uniqueness from normal lambdas isn't necessary, as
+    // these are used differently), we add 10,000 to the number.
+    // For example:
+    // _ZTSZ3foovEUlvE10005_
+    // Demangles to: typeinfo name for foo()::'lambda10005'()
+    // Note that the mangler subtracts 2, since with normal lambdas the lambda
+    // mangling number '0' is an anonymous struct mangle, and '1' is omitted.
+    // So 10,002 results in the first number being 10,000.
+    if (Ctx.IsSYCLKernelNamingDecl(ND))
+      return 10'002 + Ctx.GetSYCLKernelNamingIndex(ND);
+    return llvm::None;
+  };
+  std::unique_ptr<MangleContext> Ctx{ItaniumMangleContext::create(
+      Context, Context.getDiagnostics(), MangleCallback)};
+
+  std::string Buffer;
+  Buffer.reserve(128);
+  llvm::raw_string_ostream Out(Buffer);
+  Ctx->mangleName(GlobalDecl{VD}, Out);
+
+  if (VD->isExternallyVisible() ||
+      Context.getLangOpts().SYCLUniquePrefix.empty()) {
+    return Out.str();
+  }
+
+  return Context.getLangOpts().SYCLUniquePrefix + "___" + Out.str();
```

Comment:
Probably if we add `___` only if unique prefix is not empty, then we can make this string demanglable at least sometimes, but at the same time I'm not quite sure that this is necessary

Reply:
Can you clarify what is 'not necessary'?  Skipping the ___ in the empty-prefix case do you mean?  The thought was that it didn't really seem to help us at all...

btw, if you have a better idea than the `___` please share!

Reply:
> Can you clarify what is 'not necessary'? Skipping the ___ in the empty-prefix case do you mean?

I meant that there is a way to make the result of this built-in demanglable at least sometimes by skipping `___` in the empty-prefix case, but I'm not exactly sure that we need that functionality and therefore the skip is not necessary